pf/internal/muxer/muxer.go

```diff
@@ -119,31 +121,88 @@ func newProviderShim(provider shim.Provider) ProviderShim {
 
 // Assign each Resource and DataSource mapped in `info` whichever runtime provider defines
 // it.
-func (m *ProviderShim) ResolveDispatch(info *tfbridge.ProviderInfo) muxer.DispatchTable {
+func (m *ProviderShim) ResolveDispatch(info *tfbridge.ProviderInfo) (muxer.DispatchTable, error) {
 	var dispatch muxer.DispatchTable
 	dispatch.Resources = map[string]int{}
 	dispatch.Functions = map[string]int{}
 
-	for tfToken, res := range info.Resources {
+	unbackedResources := resolveDispatchMap(m, dispatch.Resources, info.Resources,
+		func(p shim.Provider) shim.ResourceMap { return p.ResourcesMap() })
+	unbackedDatasources := resolveDispatchMap(m, dispatch.Functions, info.DataSources,
+		func(p shim.Provider) shim.ResourceMap { return p.DataSourcesMap() })
+	joinErr := func(label string, tks []string) error {
+		return fmt.Errorf("%s without backing provider:\n- %s",
+			label, strings.Join(tks, "\n- "))
+	}
+
+	switch {
+	case len(unbackedResources) == 0 && len(unbackedDatasources) == 0:
+		return dispatch, nil
+	case len(unbackedResources) == 0:
+		return dispatch, joinErr("DataSources", unbackedDatasources)
+	case len(unbackedDatasources) == 0:
+		return dispatch, joinErr("Resources", unbackedResources)
+	default:
+		return dispatch, multierror.Append(
+			joinErr("Resources", unbackedResources),
+			joinErr("DataSources", unbackedDatasources),
+		)
+	}
+}
+
+// Resolve either resources or datasoruces into their originating providers.
+//
+// A resource/datasource is considered "from" a provider if the provider serves a
+// resource/datasource for the token or if the provider serves a resource that matches the
+// resource inserted into the global map of m. The latter logic is used to handle aliases
+// which are inserted via:
+//
+//	p.P.ResourcesMap().Set(legacyResourceName, p.P.ResourcesMap().Get(resourceName))
+func resolveDispatchMap[T interface{ GetTok() tokens.Token }](
```

Comment:
Cool trick: `type Foo interface { Bar() }` is really just saying `Foo` is a type that resolves to `interface { Bar() }`. Whenever you need to supply a type in Go, you can provider either a type name (`Foo`) or a type value `interface { Bar() }`. Even in generics.